Single inheritance alias 13784

tzeH opened the pull request:
The current aliasing mechanism makes sure, that aliases for primitve types are resolved to that basic type. It doesn't take into account, that an alias to an alias is still an alias. This PR changes that. 

The problem appeared in a 3rd party openapi spec that I had to use (https://github.com/astm-utm/Protocol/blob/cb7cf962d3a0c01b5ab12502f5f54789624977bf/utm.yaml), where they used `anyOf` with a single subtype to achieve that.

Here is the issue https://github.com/OpenAPITools/openapi-generator/issues/13784.

Basically, two methods needed to change: `DefaultCodegen.getAlias(String name)` and `DefaultCodegen.unaliasSchema(Schema schema)`. This is achieved in two ways:

1. The `typeAliases` now holds the Schema, that a type is an alias for, not just the `String`. This is necessary since `unaliasSchema` needs to return a `Schema`.
2. The `typeAliases` need to be initialized properly and include the recursion for "alias to alias" resolution. This is contained in `getAllAliases(Map<String, Schema> schemas)`.

Cleanup work: Some Tests didn't call `setOpenAPI` on the generator, which I added. The initialization of the aliases was done in `fromModel`, which I moved to `setOpenAPI`.

This merge request comes with a unit test in `DefaultCodegenTest` and a integration Test in `AbstractJavaCodegenTest`.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

tzeH commented:
@JFCote @wing328

tzeH commented:
@wing328 @jimschubert @cbornet @jaz-ah @JFCote @jmini

wing328 commented:
> Cleanup work: Some Tests didn't call setOpenAPI on the generator, which I added. The initialization of the aliases was done in fromModel, which I moved to setOpenAPI.

Can you please file a PR to fix the issue in a separate PR instead? We can get it merged a lot quicker.

wing328 commented:
> where they used anyOf with a single subtype to achieve that.

We've seen this before. If I remember correctly, it's intentional and the owner of the spec wants to inform the users the type can include other schemas in the future. Not something I would recommend. It's just another edge case that has come up before.

Not saying we need a switch for this PR but I also need to dig through other cases that may be impacted by this change.

> It doesn't take into account, that an alias to an alias is still an alias

I remember I'd a test (probably in Java fake petstore client) for that but I could be wrong.

tzeH commented:
> > Cleanup work: Some Tests didn't call setOpenAPI on the generator, which I added. The initialization of the aliases was done in fromModel, which I moved to setOpenAPI.
> 
> Can you please file a PR to fix the issue in a separate PR instead? We can get it merged a lot quicker.

Yes, see https://github.com/OpenAPITools/openapi-generator/pull/14054
